core/lib/src/rocket.rs

```diff
@@ -703,54 +703,49 @@ impl Rocket {
         Ok(self)
     }
 
-    /// Starts the application server and begins listening for and dispatching
-    /// requests to mounted routes and catchers. Unless there is an error, this
-    /// function does not return and blocks until program termination.
-    ///
-    /// # Error
-    ///
-    /// If there is a problem starting the application, a [`LaunchError`] is
-    /// returned. Note that a value of type `LaunchError` panics if dropped
-    /// without first being inspected. See the [`LaunchError`] documentation for
-    /// more information.
+    /// Similar to `launch()`, but using a custom Tokio runtime and returning
+    /// a `Future` that completes along with the server. The runtime has no
+    /// restrictions other than being Tokio-based, and can have other tasks
+    /// running on it.
     ///
     /// # Example
     ///
     /// ```rust
+    /// // This gives us the default behavior. Alternatively, we could use a
+    /// // `tokio::runtime::Builder` to configure with greater detail.
+    /// let runtime = tokio::runtime::Runtime::new();
+    ///
     /// # if false {
-    /// rocket::ignite().launch();
+    /// let fut = rocket::ignite().spawn_on(&runtime).expect("error launching server");
+    /// fut.then(|result| {
+    ///     // do things after the server shuts down
+    ///     assert!(result.is_ok());
+    /// });
     /// # }
     /// ```
-    pub fn launch(mut self) -> LaunchError {
+    pub fn spawn_on(
+        mut self,
+        runtime: &tokio::runtime::Runtime,
+    ) -> Result<impl Future<Output = Result<(), rocket_http::hyper::Error>>, LaunchError> {
```

Comment:
We shouldn't expose `rocket_http::hyper::Error`. Let's have it return `Box<dyn std::error::Error>` for now with a TODO for us to design or rename a concrete error type. I think returning `handle.err_into()` from `TryFutureExt` should work nicely to adapt the error once the type signature is changed.

Reply:
Returning a boxed error here requires the same on `.launch()`. Moving ahead with that change as a temporary solution.